お部屋のスコア結果ページを作成する

uchihiro04 opened the pull request:
## 概要 
お部屋のスコア結果ページを作成しました。
スコア結果ページでは、以下の項目をそれぞれ表示しています。
* 総合スコア（合計スコアの平均点が高い順に、２部屋を表示）
* 上位2件のスコアの比較（合計スコアの平均点が高い2部屋について、評価項目ごとのスコアの平均点をそれぞれ表示）
* お部屋ごとのスコア（各部屋の合計スコアの平均点を表示）

## スクリーンショット
* お部屋のスコア結果ページ
<img width="700" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/c709d3cd-56df-4bc2-b130-6b75c1c44861">

デベロッパツールで「iPhone SE」の画面（375 × 667）を想定して表示した場合
<img width="200" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/fd05ba83-1d6a-49fb-9ff2-51e9c28115b2">

## 関連Issue
- #49 

## 備考
以下の実装は別Issueで行います。
* デザイン
* お部屋ごとのスコアの表示形式の変更
  * 0点の部屋（スコア未入力の部屋）を除外する
  * スコアが高い順に表示する（要検討）

Close #49

uchihiro04 committed:
#49 ルーティングを設定

お部屋のスコア結果ページについて、ルーティングを設定しました。

uchihiro04 committed:
#49 HouseViewing::Scoresコントローラのひな形を作成

uchihiro04 committed:
#49 お部屋のスコア結果ページの見出しを作成

uchihiro04 committed:
#49 各項目の見出し・説明文を追加

お部屋のスコア結果ページについて、各項目に見出し・説明文を追加しました。

uchihiro04 committed:
#49 お部屋ごとのスコアを表示

お部屋ごとのスコアを表示しました。合計スコアの平均点を表示しています。表示順や0点の部屋の対応は別Issueです。

uchihiro04 committed:
#49 スコア上位の2部屋を表示

スコアが上位の2部屋の部屋名・スコアをそれぞれ表示しました。

uchihiro04 committed:
#49 スコア上位2部屋の取得方法を変更

スコア上位2部屋の取得方法を、ハッシュからインスタンスの配列に変更しました。

uchihiro04 committed:
#49 各評価項目ごとのスコアを表示

スコア上位2部屋について、各評価項目ごとのスコアを表示しました。

Yuma-Nagaoka commented on app/controllers/house_viewings/scores_controller.rb:
```diff
@@ -0,0 +1,13 @@
+      house_viewing = HouseViewing.find_by!(uuid: params[:house_viewing_uuid])
+      @rooms = house_viewing.rooms.order(:created_at)
+
+      number_of_display = 2
```
これは定数にしましょう

uchihiro04 replied:
確かに`number_of_display`（表示件数）は基本変化しないため、定数のほうが適切ですね……。

いただいたレビューをもとに変数を定数に修正しました。該当するコミットは[3216e71](https://github.com/uchihiro04/house-viewing-score-log/pull/122/commits/3216e712da0234fd0e491426322455a4f9457808)です。

Yuma-Nagaoka commented on app/views/house_viewings/scores/index.html.slim:
```diff
@@ -0,0 +1,37 @@
+      th
+      - @top_two_room_scores.each do |room_score|
+        th = room_score.name
+    - %i[living_room storage kitchen bath toilet equipment surroundings rent].each do |evaluation_item|
```
これモデルで定義した定数から取ってきたいですね

uchihiro04 replied:
モデルで定義した定数から取ってくる方が、もし評価項目が変わった場合に修正する箇所が減りますね……。

いただいたレビューをもとにRoomモデルで定義した定数から評価項目を取ってくる形に変更しました。該当するコミットは[346b545](https://github.com/uchihiro04/house-viewing-score-log/pull/122/commits/346b54559600c5157cf2258c25031f9c7b55ed99)です。

Yuma-Nagaoka commented on app/models/room.rb:
```diff
@@ -5,4 +5,23 @@ class Room < ApplicationRecord
+  def average_total_score
+    sum_total_scores = EVALUATION_ITEMS.sum { |attribute_name| scores.sum(attribute_name) }.to_f
+    if scores.length.positive?
+      (sum_total_scores / scores.length).round
```
[imho] scores.lengthってスコア入力した人の数だと思うんですが、メソッド名だけを見た時にスコアの数..？スコアの項目の数だっけ..？となったので、number_of_scorersとか、number_of_reviewers とかでメソッド定義しちゃうのもありかと思いました。
思っただけで正直このままでも全然いいです...！

uchihiro04 replied:
いただいたレビューの通り、`scores.length`は入力されたスコアの数を表しています。

コードを書いている自分は`scores.length`ですぐ理解できますが、他の方もコードを読んだときに何を指すのか把握に時間がかかる可能性は十分考えられそうです。

検討の結果、コードの可読性を上げるために`number_of_scores`メソッドを定義しました。該当するコミットは[44571f6](https://github.com/uchihiro04/house-viewing-score-log/pull/122/commits/44571f6e168453074840975aed44d846f085ea71)です。

uchihiro04 committed:
#49 変数を定数に変更

表示件数に関する変数について、基本的に数値は変更されないため定数に変更しました。

uchihiro04 committed:
#49 評価項目の配列を、定数を使用する形に変更

評価項目について、変更箇所を少なくするために定数を使用する形に変更しました。

uchihiro04 committed:
#49 スコアの数を出力するメソッドを作成

コードの可読性を考え、scores.lengthについてnumber_of_scoresメソッドとして定義しました。

Yuma-Nagaoka reviewed (approved):
LGTM

takuminasu commented on app/controllers/house_viewings/scores_controller.rb:
```diff
@@ -0,0 +1,14 @@
+
+module HouseViewings
+  class ScoresController < ApplicationController
+    NUMBER_OF_DISPLAY = 2
```
TOP_TWOとかの方がイメージしやすいですね。
NUMBER_OF_DISPLAYだととりあえず2件みたいな印象を持ちます。

uchihiro04 replied:
NUMBER_OF_DISPLAYだと、表示件数であることは分かっても2件であることに意味を持たせづらいため、TOP_TWOのほうが定数について具体的にイメージしやすい形になると、いただいたレビューから解釈しました。

レビューをもとに定数名をNUMBER_OF_DISPLAYからTOP_TWOに変更しました。該当するコミットは[92de2a1](https://github.com/uchihiro04/house-viewing-score-log/pull/122/commits/92de2a165e91915bd5ba56c6c6a368e90219b37f)です。

takuminasu commented on app/models/room.rb:
```diff
@@ -5,4 +5,29 @@ class Room < ApplicationRecord
 
   validates :name, presence: true
+
+  EVALUATION_ITEMS = %i[living_room storage kitchen bath toilet equipment surroundings rent].freeze
```
この定数はscoreの方で持ちたいですね。

uchihiro04 replied:
確かにEVALUATION_ITEMSは評価項目に関する定数なので、RoomよりScoreのほうが適切ですね……。

いただいたレビューをもとに定数EVALUATION_ITEMSをRoomモデルからScoreモデルに移動しました。該当するコミットは[6136731](https://github.com/uchihiro04/house-viewing-score-log/pull/122/commits/6136731270c8c1890921060dcc6669c124cf6f45)です。

takuminasu commented on app/models/room.rb:
```diff
@@ -5,4 +5,29 @@ class Room < ApplicationRecord
+
+  def average_total_score
+    sum_total_scores = EVALUATION_ITEMS.sum { |attribute_name| scores.sum(attribute_name) }.to_f
+    if number_of_scores.positive?
+      (sum_total_scores / number_of_scores).round
+    else
+      0
+    end
```
早期 returnしてあがた方が良さそうです。
```ruby
return 0 unless sores.present?

(sum_total_scores / sores.length).round
```

また
scores.lengthぐらいだとメソッド化すると逆に可読性が落ちる可能性があります。

uchihiro04 replied:
確かに早期returnを使った方が、スコアが未入力の場合にすぐメソッドを終了できますし、コードの行数も簡潔になることが分かりました。

また、number_of_scoresメソッドに関しましては、早期returnを使うことで使用する箇所が減ったこともあり、メソッドにしないで`scores.length`をそのまま使う形に変更しました。

該当するコミットは[07efbaf](https://github.com/uchihiro04/house-viewing-score-log/pull/122/commits/07efbaf000e1bd1064ced22ba2cd415e376ec78d)、[bde8b0f](https://github.com/uchihiro04/house-viewing-score-log/pull/122/commits/bde8b0f2edc3d429b8db71db87f22cd64a3a7c35)です。

uchihiro04 committed:
#49 定数名を変更

定数名NUMBER_OF_DISPLAYについて、表示が上位2件であることを具体的にイメージできるように名前をTOP_TWOに変更しました。

uchihiro04 committed:
#49 定数をRoomモデルからScoreモデルに移動

定数EVALUATION_ITEMSについて、評価項目に関する定数であることを考え、RoomモデルからScoreモデルに移動しました。

uchihiro04 committed:
#49 早期returnを使用する形に変更

Roomモデルのaverage_total_scoreメソッド、average_scoreメソッドについて、早期returnを使う形に変更しました。

uchihiro04 committed:
#49 number_of_scoresメソッドの削除

コードの修正により使用する箇所が少なくなったため、メソッドとして定義せず中身のコードをそのまま使用する形に変更しました。

takuminasu commented on app/models/room.rb:
```diff
@@ -5,4 +5,15 @@ class Room < ApplicationRecord
   validates :name, presence: true
+
+  def average_total_score
+    sum_total_scores = Score::EVALUATION_ITEMS.sum { |attribute_name| scores.sum(attribute_name) }.to_f
```
大量のログが吐かれてしまう...
が一旦スキップ
issueは作成した
https://github.com/uchihiro04/house-viewing-score-log/issues/125

takuminasu replied:
これは余裕があれば対応で良いです

takuminasu commented on app/views/house_viewings/scores/index.html.slim:
```diff
@@ -0,0 +1,37 @@
+        th = room_score.name
+    - Score::EVALUATION_ITEMS.each do |evaluation_item|
+      tr
+        td = t("activerecord.attributes.score.#{evaluation_item}")
```
```ruby
td = Score.human_attribute_name(evaluation_item)
```

翻訳時に `activerecord.attributes...`のようにymlに書いてある内容をベタがきするようなケースはないかと思います。

uchihiro04 replied:
`Model.human_attribute_name(attribute)`メソッドで、属性名を取得できることを失念していました……。Railsガイドなども併せて確認しました。

https://railsguides.jp/i18n.html#active-record%E3%83%A2%E3%83%87%E3%83%AB%E3%81%A7%E7%BF%BB%E8%A8%B3%E3%82%92%E8%A1%8C%E3%81%AA%E3%81%86

いただいたレビューをもとに、属性名の表示について修正を行いました。該当するコミットは[da9e82e](https://github.com/uchihiro04/house-viewing-score-log/pull/122/commits/da9e82e0c6fa0b6dbbb05ab1a21ff8378f09a82f)です。

takuminasu commented on app/views/house_viewings/scores/index.html.slim:
```diff
@@ -0,0 +1,37 @@
+  | お部屋の名前をタップ・クリックするとスコアの詳細ページに移動します。
+table
+  tbody
+    - @rooms.each do |room|
```
N+1が発生すると思います。

uchihiro04 replied:
一時的に、N+1問題が起きている箇所を通知してくれるgem（bullet）を導入してN+1問題が発生していることを確認しました。
https://github.com/flyerhzm/bullet

レビューをもとに、indexアクションの`@rooms`についてincludesメソッドを使う形に修正しました。該当するコミットは[cc61ca9](https://github.com/uchihiro04/house-viewing-score-log/pull/122/commits/cc61ca98c961ef9b48123ed495d04259bdaf0a93)です。

uchihiro04 committed:
#49 rubocopの指摘を修正

present?メソッドやガード節についてrubocopから指摘を受けたため、修正しました。

uchihiro04 committed:
#49 インスタンス名を変更

`@top_two_room_scores`について、中身がスコア上位である2つのRoomインスタンスの配列であるため、インスタンス名を`@top_two_rooms`に変更しました。

uchihiro04 committed:
#49 属性名の日本語化の方法を変更

属性名について、tメソッドを使う形からhuman_attribute_nameメソッドを使う形に変更しました。

uchihiro04 committed:
#49 N+1問題が発生するコードを修正

`@rooms`について、N+1問題が発生する形になっていたため、includesメソッドを使って修正しました。

takuminasu reviewed (approved):
LGTM
